File: src/java/cleargl/GLVector.java
package cleargl;

import java.nio.FloatBuffer;
import java.util.Arrays;

/*
@author Ulrik Günther, Loïc Royer
 */

public class GLVector
{
	protected float[] mElements;
	protected int mDimension;

	public GLVector(float... pElements)
	{
		super();
		mElements = Arrays.copyOf(pElements, pElements.length);
		mDimension = pElements.length;
	}

	public GLVector(float element, int dimension) {
		super();
		mElements = new float[dimension];
		for(int i = 0; i < dimension; i++) {
			mElements[i] = element;
		}

		mDimension = dimension;
	}

	public GLVector(GLVector pGLVector)
	{
		super();
		mElements = Arrays.copyOf(pGLVector.mElements,
															pGLVector.mElements.length);
		mDimension = pGLVector.mElements.length;
	}

	public GLVector clone()
	{
		return new GLVector(this);
	}

	public float x()
	{
		return mElements[0];
	}

	public float y()
	{
		return mElements[1];
	}

	public float z()
	{
		return mElements[2];
	}

	public float get(int pIndex)
	{
		return mElements[pIndex];
	}

	public void set(int pIndex, float pValue)
	{
		mElements[pIndex] = pValue;
	}
	
	public void plusAssign(GLVector pGLVector)
	{
		float[] lElements = pGLVector.mElements;
		for (int i = 0; i < mDimension; i++)
			mElements[i] += lElements[i];
	}

	public GLVector minus(GLVector pGLVector)
	{
		GLVector lMinus = this.clone();
		lMinus.minusAssign(pGLVector);
		return lMinus;
	}

	public GLVector plus(GLVector pGLVector)
	{
		GLVector lPlus = this.clone();
		lPlus.plusAssign(pGLVector);
		return lPlus;
	}

	public void minusAssign(GLVector pGLVector)
	{
		float[] lElements = pGLVector.mElements;
		for (int i = 0; i < mDimension; i++)
			mElements[i] -= lElements[i];
	}

	public void timesAssign(GLVector pGLVector)
	{
		float[] lElements = pGLVector.mElements;
		for (int i = 0; i < mDimension; i++)
			mElements[i] *= lElements[i];
	}

	public void divAssign(GLVector pGLVector)
	{
		float[] lElements = pGLVector.mElements;
		for (int i = 0; i < mDimension; i++)
			mElements[i] /= lElements[i];
	}

	public float times(GLVector pGLVector)
	{
		float lResult = 0;
		float[] lElements = pGLVector.mElements;
		for (int i = 0; i < mDimension; i++)
			lResult += mElements[i] * lElements[i];
		return lResult;
	}

	public float magnitude()
	{
		float lResult = 0;
		for (int i = 0; i < mDimension; i++)
		{
			float lValue = mElements[i];
			lResult += lValue * lValue;
		}
		return lResult;
	}

	public GLVector normalize()
	{
		float lFactor = 1f / magnitude();
		for (int i = 0; i < mDimension; i++)
			mElements[i] *= lFactor;
		return this;
	}

	public GLVector getNormalized()
	{
		return this.clone().normalize();
	}

	public FloatBuffer toFloatBuffer()
	{
		return FloatBuffer.wrap(mElements);
	}

	@Override
	public int hashCode()
	{
		final int prime = 31;
		int result = 1;
		result = prime * result + mDimension;
		return result;
	}

	@Override
	public boolean equals(Object obj)
	{
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		GLVector other = (GLVector) obj;
		if (mDimension != other.mDimension)
			return false;
		return true;
	}

	@Override
	public String toString()
	{
		return "GLVector [mElements=" + Arrays.toString(mElements)
						+ ", mDimension="
						+ mDimension
						+ "]";
	}

	public static GLVector getOneVector(int dimension) {
		return new GLVector(1.0f, dimension);
	}

	public static GLVector getNullVector(int dimension) {
	  return new GLVector(0.0f, dimension);
	}

}

File: src/java/cleargl/GLVector.kt
@file:JvmName("GLVector")
package cleargl

import java.nio.FloatBuffer

/**
 * Created by ulrik on 19/11/15.
 */
open public class GLVector(vararg e: Float)  {
    protected var dimension: Int = 0
    protected var elements: FloatArray? = null

    init {
        elements = e
        dimension = e.size
    }

    @Throws(Exception::class)
    fun add(v: GLVector): GLVector {
        if (this.dimension != v.dimension) {
            throw Exception("Vector dimension mismatch: ${this.dimension} vs. ${v.dimension}")
        }

        val n = GLVector()
        val e = FloatArray(v.dimension)
        n.set(e)

        for (i in 0..this.dimension - 1) {
            n.setElement(i, this.getElement(i) + v.getElement(i))
        }

        return n
    }

    @Throws(Exception::class)
    fun scalarProduct(v: GLVector): Float {
        if (this.dimension != v.dimension) {
            throw Exception("Vector dimension mismatch!")
        }

        var value = 0.0f

        for (i in 0..this.dimension - 1) {
            value += this.getElement(i) * v.getElement(i)
        }

        return value
    }

    fun mult(num: Float): GLVector {
        val n = GLVector()

        for (i in 0..this.dimension - 1) {
            n.setElement(i, this.getElement(i) * num)
        }

        return n
    }

    fun mult(num: Int): GLVector {
        return mult(num.toFloat())
    }

    fun getElement(index: Int): Float {
        return this.elements!![index]
    }

    fun set(elements: FloatArray) {
        this.elements = elements
    }

    fun setElement(index: Int, value: Float) {
        this.elements!![index] = value
    }

    fun x(): Float {
        return this.getElement(0)
    }

    fun y(): Float {
        return this.getElement(1)
    }

    fun z(): Float {
        return this.getElement(2)
    }

    fun w(): Float {
        return this.getElement(3)
    }

    fun toFloatBuffer(): FloatBuffer {
        return FloatBuffer.wrap(elements)
    }

    fun magnitude(): Float =
        Math.sqrt(elements!!.map { e -> Math.pow(e.toDouble(), 2.0)}.sum()).toFloat()

    fun normalize() {
        elements = elements!!.map { e -> e/magnitude() }.toFloatArray()
    }

    fun getNormalized(): GLVector {
        var n: GLVector = GLVector(0.0f, 0.0f, 0.0f)
        n.set(elements!!.map { e -> e/magnitude() }.toFloatArray())
        return n
    }

    operator fun plus(b: GLVector): GLVector {
        return add(b)
    }

    operator fun minus(b: GLVector): GLVector {
        return add(b.mult(-1.0f))
    }

    override fun toString(): String {
        return "(${elements!![0]}, ${elements!![2]}, ${elements!![1]})"
    }
}
